src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -635,6 +635,8 @@ bool QgsSymbolLegendNode::setData( const QVariant &value, int role )
 
   vlayer->renderer()->checkLegendSymbolItem( mItem.ruleKey(), value == Qt::Checked );
 
+  QgsProject::instance()->setDirty( true );
```

Comment:
We should never reference QgsProject::instance() in new core code. 
```suggestion
  if ( QgsProject* project = vlayer->project() )
     project->setDirty( true );
```

Reply:
Right, I've commited the suggestion and will keep that in mind!